### ISSUE TITLE
migrate kubeadm control-plane taint

### ISSUE THREAD

**Issue by BenTheElder** [kubernetes-sigs/kind#1699]
See: https://github.com/kubernetes/kubeadm/issues/2200

We'll need to keep in sync with the planning there and ensure kind's handling of this does not block work there.

Currently off the top of my head we only reference the `node-role.kubernetes.io/master:NoSchedule` taint to remove it from single-node clusters. We'll want to handle removing the correct taint based on version.

**Comment by BenTheElder** [comment]
This is blocked on upstream changes, probably in Kubernetes 1.20?
When the upstream project is ready we'll coordinate on this.
Frozen to avoid fejta-bot noise / accidental closure before then.

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/pull/95382

**Comment by BenTheElder** [comment]
Updates:
- wg-naming was concluded
- I think this is closing out in kubeadm in 1.24

There are still various bits in this repo referencing the old label, we'll need to carefully handle new + old depending on version.

Tentatively flagging as a blocker for kind v0.13

**Comment by BenTheElder** [comment]
In 1.24:

> [*] continue the rename of the kubeadm "master" label / taint to
"control-plane".
in 1.24 we are adding a new taint
"node-role.kubernetes.io/control-plane" with effect "NoSchedule" and
removing the "node-role.kubernetes.io/master" label on nodes. in 1.20
we instructed all users to preemptively add toleration for this taint
in their workloads and to start using the newly added label
"node-role.kubernetes.io/control-plane".
this plan is detailed in the KEP linked here:
https://github.com/kubernetes/kubeadm/issues/2200

See: https://groups.google.com/d/msgid/kubernetes-sig-cluster-lifecycle/CAGDbWi_5tQ83H_bKYKn2G1O_UWQviUsHmnspfboB1%3DHZ-tVv7A%40mail.gmail.com

We will need to handle removing the taint for single node clusters, and we will need to make sure waitForReady handles the updated label as well. Docs for workload samples pinning to the control plane node may also need updating (I _think_ we already used our own label everywhere for that but ...)

**Comment by neolit123** [comment]
PR with the label / taint adjustments for 1.24 is here:
https://github.com/kubernetes-sigs/kind/issues/1699

**Comment by BenTheElder** [comment]
the kind changes are now done thanks to @neolit123, but we still need to review the docs / samples

**Comment by wolf99** [comment]
Having searched for incidences of "master" across the repo, it seems like all occasions are either 

1. Accounted for by:
    - https://github.com/kubernetes-sigs/kind/pull/2595
    - https://github.com/kubernetes-sigs/kind/pull/2603
2. Links to files on the master branch in the GitHub repo structure
   - As it seems like this issue is specifically about renaming the taint rather than complete removal of the term "master" then renaming the master branch to remove the term from the links appears to be out of scope

Thus it looks like this issue could be closed?

**Comment by BenTheElder** [comment]
Thanks for checking those!

We probably still at least need to check the external manifests we reference for install etc. (e.g. ingress-nginx).

**Comment by neolit123** [comment]
Note that https://github.com/kubernetes-sigs/kind/pull/2595 left some todos for 1.15 linking to this issue.

**Comment by wolf99** [comment]
@neolit123 just to confirm my understanding, it seemed like such todos are deprecations, i.e. things to be removed in the future.

Did I understand correctly, or do you mean there are parts that should be removed right now?

**Comment by BenTheElder** [comment]
It was important that we handle at least the Kubernetes 1.24 parts for this kind release given how the timing has wound up, but what @neolit123 is referring to I believe is 1.25 changes that have not happened yet (see this particular comment on the linked PR https://github.com/kubernetes-sigs/kind/pull/2595#discussion_r788825554)

Bumping the remainder of this to the next milestone, as aojea has a fix for the other issue.

**Comment by cfryanr** [comment]
Hi folks,

I apologize if I am making this comment in the wrong place. I was considering creating a new bug issue but then I saw that this issue is still open. Please let me know if there is a better place to post this.

I'm using kind version v0.11.1. I'm also using a node image that I built using `kind build node-image` with the `--kube-root` as the HEAD of https://github.com/kubernetes/kubernetes.git.

When I create a single-node kind cluster, my cluster shows:

```
% kubectl get nodes
NAME                 STATUS   ROLES           AGE    VERSION
kind-control-plane   Ready    control-plane   101m   v1.24.0-alpha.3.139+267272efe0725e

% kubectl get node -o jsonpath='{.spec.taints}'
[{"effect":"NoSchedule","key":"node-role.kubernetes.io/control-plane"}]
```

Shouldn't that taint be automatically removed by [this code](https://github.com/kubernetes-sigs/kind/blob/bf63502ed1dfdee6c898ee5cc4318b89f0e09980/pkg/cluster/internal/create/actions/kubeadminit/init.go#L124-L126) since it is a single-node cluster? Is it failing to remove the taint because `v1.24.0-alpha.3.139+267272efe0725e` is less than `v1.24.0`, since it is a pre-release? Is it a bug?

This seems very related to the PR referenced by @BenTheElder above.

Having the `control-plane` taint prevents all my Deployments from deploying on these clusters, since my Deployments do not tolerate that taint. For my single-node kind clusters which use older node images (1.23, 1.22, etc.), the node has no taint, as expected.

**Comment by aojea** [comment]
> Shouldn't that taint be automatically removed by [this code](https://github.com/kubernetes-sigs/kind/blob/bf63502ed1dfdee6c898ee5cc4318b89f0e09980/pkg/cluster/internal/create/actions/kubeadminit/init.go#L124-L126) since it is a single-node cluster? Is it failing to remove the taint because `v1.24.0-alpha.3.139+267272efe0725e` is less than `v1.24.0`, since it is a pre-release? Is it a bug?

that code is in a recent version of kind, you are using 0.11.1

**Comment by neolit123** [comment]
> Shouldn't that taint be automatically removed by [this code](https://github.com/kubernetes-sigs/kind/blob/bf63502ed1dfdee6c898ee5cc4318b89f0e09980/pkg/cluster/internal/create/actions/kubeadminit/init.go#L124-L126) since it is a single-node cluster? Is it failing to remove the taint because v1.24.0-alpha.3.139+267272efe0725e is less than v1.24.0, since it is a pre-release? Is it a bug?

with pre-releases in mind, a better check would be to `if kubeVersion.Major() == 1 && kubeVersion.Minor() < 24 { /* remove old taint only */ }` instead of the current `if kubeVersion.LessThan(version.MustParseSemantic("v1.24.0")) { ... }`

a PR with a fix is welcome.

**Comment by BenTheElder** [comment]
> I'm using kind version v0.11.1. I'm also using a node image that I built using kind build node-image with the --kube-root as the HEAD of https://github.com/kubernetes/kubernetes.git.

Unfortunately any given release version may not work with _HEAD_ of Kubernetes due to changes like this, for developing Kubernetes @ HEAD we recommend kind @ HEAD (which should be stable for this purpose, it is what Kubernetes uses to test, and we run Kubernetes CI prior to merges). 

At HEAD of kind HEAD of Kubernetes should work, generally.

---

> with pre-releases in mind, a better check would be to if kubeVersion.Major() == 1 && kubeVersion.Minor() < 24 { ... } instead of the current if kubeVersion.LessThan(version.MustParseSemantic("v1.24.0")) { ... }




We have also in the past used the *exact* version where it becomes true, if you find the kubernetes commit with the change you can get the version from the build tools and use that in kind, the versions are sortable (they include commits since last tag etc. xref #2618)

hack/print-workspace-status.sh in Kubernetes will print the version.

**Comment by cfryanr** [comment]
Thanks, everyone. Those were all useful suggestions regarding my comment above.

**Comment by BenTheElder** [comment]
one last followup, then I think we're done https://github.com/kubernetes-sigs/kind/pull/2779

though xref: https://github.com/kubernetes/ingress-nginx/issues/8605 external project may need updates
